src/form/middleware.ts

```diff
@@ -0,0 +1,192 @@
+import { create } from '@dojo/framework/core/vdom';
+import cache from '@dojo/framework/core/middleware/cache';
+import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
+
+export type Validity = true | { valid: false | undefined; message: string };
+
+interface Callbacks<S> {
+	onValue?: (values: Partial<S>) => void;
+}
+
+export interface FormMiddleware<S> {
+	value: {
+		(): Partial<S>;
+		(values: S): S;
+		(values: Partial<S>): Partial<S>;
+	};
+	submit: {
+		(callback: (values: Partial<S>) => void): void;
+		(callback: (values: S) => void, defaults: S): void;
+	};
+	onValue(callback: (values: Partial<S>) => void): void;
+	disabled(disabled?: boolean): boolean;
+	valid(): boolean;
+	reset(): void;
+	field<K extends keyof S>(name: K, required?: boolean): Field<S, K>;
+}
+
+export interface Field<S, K extends keyof S> {
+	value: {
+		(newValue: S[K]): S[K];
+		(): S[K] | undefined;
+	};
+	valid(valid?: boolean, message?: string): Validity;
+	required(required?: boolean): boolean;
+	disabled(required?: boolean): boolean;
+}
+
+function compareValidity(a: Validity, b: Validity) {
+	if (typeof a === 'boolean' && typeof b === 'boolean') {
+		return true;
+	}
+	if (typeof a === 'boolean' || typeof b === 'boolean') {
+		return false;
+	}
+	if (b.valid === undefined) {
+		return false;
+	}
+	return a.message === b.message;
+}
+
+interface State<S> {
+	required: Record<string, boolean>;
+	values: Partial<S>;
+	valid: Record<string, Validity>;
+	disabled: Record<string, boolean>;
+	formDisabled: boolean;
+}
+
+export const createFormMiddleware = <S extends Record<string, any> = any>() => {
+	const icache = createICacheMiddleware<State<S>>();
+	const factory = create({ icache, cache });
+
+	const formMiddleware = factory(function Form({
+		middleware: { icache, cache }
+	}): FormMiddleware<S> {
+		return {
```

Comment:
Can we include a dirty state? It would be nice to use this to tell if the current values don't match the initial values. This would be used for "you have unsaved changes" confirmations and such.

Reply:
perhaps this is something you could raise an issue for and we could do as a follow up?

Reply:
Created https://github.com/dojo/widgets/issues/954